bcdaworker/worker/alr.go

```diff
@@ -0,0 +1,215 @@
+	}
+
+	err := conf.Checkout(worker) // worker is already a reference, no & needed
+	if err != nil {
+		logrus.Fatal("Could not get data from conf for ALR.", err)
+	}
+
+	return worker
+}
+
+func (a *alrWorker) ProcessAlrJob(
+	ctx context.Context,
+	jobArgs models.JobAlrEnqueueArgs,
+) error {
+
+	// Parse the jobAlrEnqueueArgs
+	aco := jobArgs.CMSID
+	id := jobArgs.ID
+	MBIs := jobArgs.MBIs
+	lowerBound := jobArgs.LowerBound
+	upperBound := jobArgs.UpperBound
+
+	// Pull the data from ALR tables (alr & alr_meta)
+	alrModels, err := a.GetAlr(ctx, aco, MBIs, lowerBound, upperBound)
+	if err != nil {
+		logrus.Error(err)
+		return err
+	}
+
+	// Set up IO operation to dump ndjson
+	fileUUID := uuid.New()
+	f, err := os.Create(fmt.Sprintf("%s/%d/%s.ndjson", a.FHIR_STAGING_DIR,
+		id, fileUUID))
+	if err != nil {
+		logrus.Error(err)
+		return err
+	}
+	w := bufio.NewWriter(f)
+	defer utils.CloseFileAndLogError(f)
+
+	// Serialize data into JSON
+	marshaller, err := jsonformat.NewPrettyMarshaller(jsonformat.STU3)
+	if err != nil {
+		logrus.Error(err)
+		return err
+	}
+
+	// data is what will be sent over the channel to the go routine
+	type data struct {
+		patient      *resources_go_proto.Patient
+		observations []*resources_go_proto.Observation
+	}
+
+	// Creating channels for go routine
+	// c is buffered b/c IO operation is slower than unmarshalling
+	c := make(chan data, 1000) // 1000 rows before blocking
+	result := make(chan error)
+
+	// Go rountine + closure that will write to file
+	// Two channels involved with this closure. One receiving "data", and the
+	// other will send an error is there was any error. The worker thread will
+	// will also use the result channel to wait for the IO to finish.
+	go func(c chan data, w *bufio.Writer, result chan error) {
+		// Receive data through channel
+
+		for i := range c {
+			// marshall
+			patientb, err := marshaller.Marshal(i.patient)
+			patients := string(patientb) + "\n"
+			if err != nil {
+				// Make sure to send err back to the other thread
+				result <- err
+				return
+			}
+			var observations []string
+
+			for _, observation := range i.observations {
+				obsMarshalled, err := marshaller.Marshal(observation)
+				if err != nil {
+					result <- err
+					return
+				}
+				observations = append(observations, string(obsMarshalled))
+			}
+
+			observation := strings.Join(observations, "\n")
+
+			// IO operation
+			_, err = w.WriteString(patients + observation)
+			if err != nil {
+				result <- err
+				return
+			}
+			err = w.Flush()
+			if err != nil {
+				result <- err
+				return
+			}
+		}
+
+		// Add the location of the ndjson
+		a.ndjsonFilename = fileUUID
+
+		// Everything went a-ok... send nil
+		result <- nil
+
+	}(c, w, result)
+
+	// Marshall into JSON and send it over the channel
+	for i := range alrModels {
+		patient, observations := alr.ToFHIR(&alrModels[i], alrModels[i].Timestamp)
+		c <- data{patient, observations}
```

Comment:
I'm wondering if we need to use the above goroutine and channel here. I'm thinking if we had a process that incrementally reuturn the postgres results (like a paging mechanism), then we would be able to see a lot of impacts by keeping the db IO and file IO separate, but by the time you get here, you've finished all of the db IO, and are now doing all of the file IO operations.